include/ensmallen_bits/callbacks/progress_bar.hpp

```diff
@@ -164,8 +164,43 @@ class ProgressBar
     output << "] " << progress << "% - ETA: " << (size_t) stepTimer.toc() *
         (epochSize - step + 1) % 60 << "s - loss: " <<
         objective / (double) step <<  "\r";
-    output.flush();
 
+    if (optimizer.MaxIterations() < function.NumFunctions() &&
+        optimizer.MaxIterations() != 0)
+    {
+      if (progress == (size_t)((double)(optimizer.MaxIterations()) /
+                                (double)(function.NumFunctions()) * 100) &&
+                      (int)(steps * optimizer.BatchSize() -
+                      optimizer.MaxIterations()) >= 0)
+      {
+        const size_t progress = ((double) step / epochSize) * 100;
+        output << step++ << "/" << epochSize << " [";
+        for (size_t i = 0; i < 100; i += width)
+        {
+          if (i < progress)
+          {
+            output << "=";
+          }
+          else if (i == progress)
+          {
+            output << "|";
+          }
+          else
+          {
+            output << "x";
+          }
+        }
+
+        output << "] " << progress << "% - ETA: " << (size_t) stepTimer.toc() *
+            (epochSize - step + 1) % 60 << "s - loss: " <<
+            objective / (double) step <<  "\r";
+        output << "\n" << "Optimization finished before the end of an epoch "
+                       << "because of the entire dataset not being passed to "
+                       << "the optimizer." << "\n" << "\r";
```

Comment:
Do you think that we could improve this error message?  To a user I don't think it's clear what's going on.  I think maybe something more like this could be clearer:

```
Optimization terminated; maximum iterations reached before the end of an epoch.
```

Also, why the `\r` at the end of the line?  `\n` or `std::endl` should be sufficient.  I don't imagine that we are printing anything *after* termination such that the line needs to be rolled back.